### PR TITLE
feature: storage_metadata_write_full_object_key

### PR DESCRIPTION
**Pull request by CheSema**
This is the first step in context of that issue https://github.com/ClickHouse/ClickHouse/issues/55217
The second step is https://github.com/ClickHouse/ClickHouse/pull/55614

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce the feature `storage_metadata_write_full_object_key`. If it is set as `true` then metadata files are written with new format VERSION_FULL_OBJECT_KEY. With that format CH stores full remote object key in the metadata file.

Todo
- add test `S3PlainObjectStorage` + `storage_metadata_write_full_object_key = true`
- add test `MetadataStorageFromStaticFilesWebServer` + `storage_metadata_write_full_object_key = true`

There no implementation for move operation in `S3PlainObjectStorage` and `MetadataStorageFromStaticFilesWebServer`. As a result neither `Log` nor `MergeTree` tables work on that types of storages.
Instead this tests I did other thing. I turned new flag on here https://github.com/ClickHouse/ClickHouse/pull/56027 and checked the tests.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit d30d704d92c7086cd4b060cdf47c6316e76d2e5e with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55566/d30d704d92c7086cd4b060cdf47c6316e76d2e5e/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by CheSema** [comment]
The code is complete
There are only several tests left, I dont expect that code is going to change a lot while Im writing them.
So, PR is opened for review.

**Comment by CheSema** [comment]
Stress test (msan) — Sanitizer assert
https://github.com/ClickHouse/ClickHouse/issues/55148#issuecomment-1773399093

**Comment by CheSema** [comment]
Upgrade check tsan -- https://github.com/ClickHouse/ClickHouse/issues/49156

File: src/Disks/ObjectStorages/DiskObjectStorageMetadata.cpp [hunk 1]
```diff
@@ -73,29 +88,63 @@ void DiskObjectStorageMetadata::deserialize(ReadBuffer & buf)
...
 {
     ReadBufferFromString buf(data);
     deserialize(buf);
 }
 
 void DiskObjectStorageMetadata::serialize(WriteBuffer & buf, bool sync) const
 {
-    writeIntText(VERSION_INLINE_DATA, buf);
+    /// There are the changes for backward compatibility
+    /// No new file should be write as VERSION_FULL_OBJECT_KEY until storage_metadata_write_full_object_key feature is enabled
```
Comment (antonio2368) [inline src/Disks/ObjectStorages/DiskObjectStorageMetadata.cpp:100]:
```suggestion
    /// No new file should be written as VERSION_FULL_OBJECT_KEY until storage_metadata_write_full_object_key feature is enabled
```

File: src/Disks/ObjectStorages/DiskObjectStorageMetadata.cpp [hunk 2]
```diff
@@ -73,29 +88,63 @@ void DiskObjectStorageMetadata::deserialize(ReadBuffer & buf)
...
+void DiskObjectStorageMetadata::deserializeFromString(const String & data)
 {
     ReadBufferFromString buf(data);
     deserialize(buf);
 }
 
 void DiskObjectStorageMetadata::serialize(WriteBuffer & buf, bool sync) const
 {
-    writeIntText(VERSION_INLINE_DATA, buf);
+    /// There are the changes for backward compatibility
```
Comment (antonio2368) [inline src/Disks/ObjectStorages/DiskObjectStorageMetadata.cpp:99]:
```suggestion
    /// These are the changes for backward compatibility
```

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp [hunk 3]
```diff
@@ -663,46 +664,53 @@ std::unique_ptr<WriteBufferFromFileBase> DiskObjectStorageTransaction::writeFile
...
 {
-    String blob_name;
+    auto object_key = object_storage.generateObjectKeyForPath(path);
     std::optional<ObjectAttributes> object_attributes;
 
-    blob_name = object_storage.generateBlobNameForPath(path);
     if (metadata_helper)
     {
+        if (!object_key.hasPrefix())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "metadata helper is not supported with abs paths");
```
Comment (antonio2368) [inline src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:673]:
```suggestion
            throw Exception(ErrorCodes::LOGICAL_ERROR, "metadata helper is not supported with absolute paths");
```

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp [hunk 4]
```diff
@@ -663,46 +664,53 @@ std::unique_ptr<WriteBufferFromFileBase> DiskObjectStorageTransaction::writeFile
...
-    auto object = StoredObject(fs::path(metadata_storage.getObjectStorageRootPath()) / blob_name);
-    auto write_operation = std::make_unique<WriteFileObjectStorageOperation>(object_storage, metadata_storage, object);
+    /// seems ok
+    auto object = StoredObject(object_key.serialize());
     std::function<void(size_t count)> create_metadata_callback;
 
     if (autocommit)
     {
-        create_metadata_callback = [tx = shared_from_this(), mode, path, blob_name](size_t count)
+        create_metadata_callback = [tx = shared_from_this(), mode, path, object_key](size_t count)
```
Comment (antonio2368) [inline src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:692]: move `object_key` here?

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp [hunk 5]
```diff
@@ -663,46 +664,53 @@ std::unique_ptr<WriteBufferFromFileBase> DiskObjectStorageTransaction::writeFile
...
 
             tx->metadata_transaction->commit();
         };
     }
     else
     {
-        create_metadata_callback = [object_storage_tx = shared_from_this(), write_op = write_operation.get(), mode, path, blob_name](size_t count)
+        auto write_operation = std::make_unique<WriteFileObjectStorageOperation>(object_storage, metadata_storage, object);
+
+        create_metadata_callback = [object_storage_tx = shared_from_this(), write_op = write_operation.get(), mode, path, object_key](size_t count)
```
Comment (antonio2368) [inline src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:713]: move `object_key`

File: src/Common/ObjectStorageKey.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    object_key.is_relative = true;
+    return object_key;
+}
+
+ObjectStorageKey ObjectStorageKey::createAsAbsolute(String key_)
+{
+    ObjectStorageKey object_key;
+    object_key.key = std::move(key_);
+    object_key.is_relative = true;
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/ObjectStorageKey.cpp:65]: You are right here. This is essential.

File: src/Common/ObjectStorageKey.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "object key has no suffix, key: {}", key);
+    return suffix;
+}
+
+const String & ObjectStorageKey::serialize() const
+{
+    return key;
+}
+
+ObjectStorageKey ObjectStorageKey::createAsRelativeAnyway(String key_)
```
[the remark this answers is not in the document]
Comment (CheSema) [reply src/Common/ObjectStorageKey.cpp:37]: I wanted to emphasize here, that despite only one part of the key has been passed, the result is relative key with empty prefix.

However if follow set of methods looks detailed enough without `Anyway`, lets get rid off of that emphasizing.
```
ObjectStorageKey::createAsAbsolute(String key_)
ObjectStorageKey::createAsRelative(String key_)
ObjectStorageKey::createAsRelative(String prefix_, String suffix_)
```

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp (hunk 4 above)
Comment (CheSema) [reply src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:692]: Yes, I will add moving here.

File: src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp (hunk 5 above)
Comment (CheSema) [reply src/Disks/ObjectStorages/DiskObjectStorageTransaction.cpp:713]: and here

File: src/Storages/System/StorageSystemRemoteDataPaths.cpp [hunk 8]
```diff
@@ -78,7 +78,9 @@ Pipe StorageSystemRemoteDataPaths::read(
                     col_local_path->insert(local_path);
                     col_remote_path->insert(object.remote_path);
                     col_size->insert(object.bytes_size);
-                    col_namespace->insert(common_prefox_for_objects);
+
+                    col_namespace->insertDefault();
+                    //col_namespace->insert(common_prefox_for_objects);
```
Comment (antonio2368) [inline src/Storages/System/StorageSystemRemoteDataPaths.cpp:83]: Delete this line?

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**